compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELinalgFunctionOutlining.cpp

```diff
@@ -0,0 +1,115 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree-amd-aie/IR/AMDAIEOps.h"
+#include "iree-amd-aie/Transforms/AMDAIEUtils.h"
+#include "iree-amd-aie/Transforms/Passes.h"
+#include "mlir/Conversion/FuncToLLVM/ConvertFuncToLLVM.h"
+#include "mlir/Conversion/FuncToLLVM/ConvertFuncToLLVMPass.h"
+#include "mlir/Dialect/Linalg/IR/LinalgInterfaces.h"
+#include "mlir/Dialect/Linalg/Transforms/Hoisting.h"
+#include "mlir/Dialect/Linalg/Transforms/Transforms.h"
+#include "mlir/Dialect/Vector/IR/VectorOps.h"
+
+#define DEBUG_TYPE "iree-amdaie-linalg-function-outlining"
+
+namespace mlir::iree_compiler::AMDAIE {
+
+namespace {
+
+/// Utility to outline the linalg compute op.
+static FailureOr<func::FuncOp> outlinedToAFunction(
+    IRRewriter &rewriter, ModuleOp &moduleOp, linalg::LinalgOp &computeOp) {
+  // Form outlined FuncName.
+  std::string computeName = "";
+  if (isMatmul(computeOp)) {
+    computeName = "_matmul";
+  } else if (isElementwise(computeOp)) {
+    computeName = "_elementwise";
+  } else {
+    return failure();
+  }
+  std::string outlinedFuncName =
+      computeOp->getName().stripDialect().str() + computeName + "_outlined";
+  if (auto outlinedFuncOp = dyn_cast_if_present<func::FuncOp>(
+          moduleOp.lookupSymbol(outlinedFuncName))) {
```

Comment:
Hi @newling - thank you so much for the nudge in this case!

I initially thought that adding just a clear demarcation using `isMatmul` and `isElementwise` would be general enough for a particular dispatch, but missed this point in the generalisation.

I've therefore addressed this comment by indeed going ahead with `Option 1` as you suggested - involved playing around with IREE's `isStructurallyEquivalentTo` but I couldn't make that to work, therefore went ahead with upstream's `isEquivalentTo`.

Please take a look. I've also added corresponding comments in the codebase.

Reply:
Nice, I didn't know about `isEquivalentTo`. This seems to do exactly what we want. 

I probably would have tried a slightly different design -- I would have created a `func.func` for every compute operation, and then compared func.funcs for equivalence (using `isReqionEquivalentTo`) before returning from outlinedToAFunc. This would require keeping a `DenseSet<FuncOp>` instead of the DenseMap, which would mean we could erase compute ops immediately rather than waiting till the end of the pass (smaller mental stack while reading code!) 

This is just a minor design preference though -- I will make a PR with this change after this lands.

Reply:
> I would have created a func.func for every compute operation

Hi @newling - please refer to [this](https://github.com/nod-ai/iree-amd-aie/pull/862#discussion_r1827565532) thread. I strongly feel that we shouldn't be creating `func.func` for every compute ops if we could just reuse it by looking up the Symbol table.